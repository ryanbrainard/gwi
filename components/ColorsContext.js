import React from "react";

export default (ColorsContext = React.createContext());

export function provideColors(colors, Component) {
  return props => {
    return (
      <ColorsContext.Provider value={colors}>
        <Component {...props} />
      </ColorsContext.Provider>
    );
  };
}

export function withColors(Component) {
  return props => {
    return (
      <ColorsContext.Consumer>
        {colors => <Component {...props} colors={colors} />}
      </ColorsContext.Consumer>
    );
  };
}
