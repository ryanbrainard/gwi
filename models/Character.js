import VoicePlayerMRU from "./VoicePlayerMRU";

export default class Character {
  static _chars = {};

  static register(char) {
    this._chars[char.name] = char;
    return char;
  }

  static find(name) {
    return this._chars[name];
  }

  constructor(name, group, voices) {
    this._name = name;
    this._group = group;
    this._voices = voices;

    this.player = new VoicePlayerMRU(this._voices);
  }

  get key() {
    return this.name;
  }

  get name() {
    return this._name;
  }

  get group() {
    return this._group.map(c => Character.find(c));
  }

  async preloadPlay() {
    await this.player.load();
  }

  async play() {
    await this.player.play();
  }
}
