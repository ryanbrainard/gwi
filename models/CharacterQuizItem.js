export default class CharacterQuizItem {
  constructor(char, stateSetter) {
    this._char = char;
    this._stateSetter = stateSetter;
    this._success = false;
  }

  get character() {
    return this._char;
  }

  get success() {
    return this._success;
  }

  set success(value) {
    this._success = value;
    this._stateSetter(this); // TODO: good idea or just gross?
  }
}
