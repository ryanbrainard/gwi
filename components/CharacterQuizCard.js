import { Feather } from "@expo/vector-icons";
import _ from "lodash";
import PropTypes from "prop-types";
import React from "react";
import { Button, StyleSheet, View } from "react-native";
import Character from "../models/Character";
import CharacterTile from "./CharacterTile";
import Settings from "../models/Settings";

export default class CharacterQuizCard extends React.Component {
  static propTypes = {
    char: PropTypes.instanceOf(Character).isRequired,
    universe: PropTypes.arrayOf(PropTypes.instanceOf(Character)).isRequired,
    gotoNext: PropTypes.func,
    numChoices: PropTypes.number,
    _shuffle: PropTypes.func
  };

  static defaultProps = {
    numChoices: 3,
    _shuffle: _.shuffle
  };

  constructor(props) {
    super(props);

    this.state = {
      charChoices: this.generateChoices(),
      success: undefined
    };
  }

  generateChoices() {
    const { char, universe, numChoices, _shuffle } = this.props;

    const others = universe.filter(cu => cu !== char);
    const shuffledOthers = _shuffle(others);
    const enoughShuffledOthers = shuffledOthers.slice(0, numChoices - 1);
    const choices = enoughShuffledOthers.concat(char);
    const shuffledChoices = _shuffle(choices);

    return shuffledChoices;
  }

  componentDidUpdate(_, prevState) {
    if (this.props.gotoNext && this.state.success && !prevState.success) {
      Settings.get(Settings.KEYS.ADVANCE_ON_SUCCESS).then(
        setting => setting && this.props.gotoNext()
      );
    }
  }

  render() {
    const { char, gotoNext } = this.props;
    const { charChoices, success } = this.state;

    return (
      <View style={styles.container}>
        <CharacterTile char={char} show={!!success} />

        {charChoices.map(cc => {
          const ccSuccess = cc === char;
          return (
            <Button
              key={cc.key}
              title={cc.name}
              onPress={() => {
                if (!ccSuccess) {
                  cc.play();
                }
                this.setState({ success: ccSuccess });
              }}
            />
          );
        })}

        {this.renderSuccess(success)}
      </View>
    );
  }

  renderSuccess(success) {
    switch (success) {
      case null:
        return null;
      case true:
        return <Feather name="check-circle" size={48} color="#3F681C" />;
      case false:
        return <Feather name="x-circle" size={48} color="tomato" />;
    }
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "white",
    alignItems: "center"
  }
});
