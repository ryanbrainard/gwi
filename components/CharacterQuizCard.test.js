import React from "react";
import renderer from "react-test-renderer";
import Character from "../models/Character";
import CharacterQuizItem from "../models/CharacterQuizItem";
import CharacterQuizCard from "./CharacterQuizCard";
import { provideColors } from "./ColorsContext";

const _CharacterQuizCard = provideColors(
  { primary: "#FFA500" },
  CharacterQuizCard
);

describe("renders correctly", () => {
  const group = ["A", "B", "C", "D"];
  group.forEach(char => Character.register(new Character(char, group)));
  const correctChar = Character.find("C");
  const incorrectChar = Character.find("D");
  const stateSetter = jest.fn();
  const newItem = tap => {
    const item = new CharacterQuizItem(
      correctChar,
      stateSetter,
      () => "00000000-0000-0000-0000-000000000000",
      s => s
    );
    if (tap != null) {
      tap(item);
    }
    return item;
  };

  const tests = {
    new: newItem(),
    success: newItem(item => (item.answered = correctChar)),
    fail: newItem(item => (item.answered = incorrectChar))
  };

  Object.entries(tests).forEach(([testName, testItem]) => {
    test(testName, () => {
      const tree = renderer
        .create(<_CharacterQuizCard charItem={testItem} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
