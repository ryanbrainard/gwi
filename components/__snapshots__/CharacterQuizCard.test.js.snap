// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "flex": 1,
      },
      Object {
        "padding": 75,
      },
    ]
  }
>
  <View
    style={
      Object {
        "borderColor": "#FFFFFF",
        "borderWidth": 2,
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFA500",
          "height": 225,
          "justifyContent": "center",
          "opacity": 1,
          "width": 225,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000000",
            "fontSize": 24,
            "fontWeight": "bold",
          }
        }
      >
        ?
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "margin": 10,
        },
        Object {
          "flexDirection": "column",
        },
      ]
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#A9A9A9",
          "borderRadius": 150,
          "height": 75,
          "justifyContent": "center",
          "margin": 30,
          "opacity": 1,
          "width": 150,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#000000",
              "fontWeight": "bold",
            },
            Object {
              "fontSize": 30,
            },
          ]
        }
      >
        A
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#A9A9A9",
          "borderRadius": 150,
          "height": 75,
          "justifyContent": "center",
          "margin": 30,
          "opacity": 1,
          "width": 150,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#000000",
              "fontWeight": "bold",
            },
            Object {
              "fontSize": 30,
            },
          ]
        }
      >
        B
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#A9A9A9",
          "borderRadius": 150,
          "height": 75,
          "justifyContent": "center",
          "margin": 30,
          "opacity": 1,
          "width": 150,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#000000",
              "fontWeight": "bold",
            },
            Object {
              "fontSize": 30,
            },
          ]
        }
      >
        C
      </Text>
    </View>
  </View>
</View>
`;
